### PR TITLE
System integration

### PR DESCRIPTION
**Pull request by blacksph3re**
Added logger plug which was prepared by Sergey
Added option to authenticate through a system-wide supertoken which can be supplied by secret. For that, set the environment variable SUPERTOKEN_FILE=/var/run/secrets/supertoken

### CODE REVIEW

**Comment by serge1peshcoff** [comment]
Cool, thanks! I've left a few comments on that.

One question: how do you do requests with supertoken, by providing the `X-Auth-Token` as with regular requests?

**Comment by blacksph3re** [comment]
I don't see your comments, where did you put them?

To your question: Yes exactly, you provide it and you don't need to log in to anything anymore. So we should really keep this thing secret, or disable it altogether in prod. Disable -> Set nil as supertoken

**Comment by serge1peshcoff** [comment]
@blacksph3re for the comments, check above ;)
basically there were a few things: remove my io.inspect() logs and add request body for PUT and POST requests like I did for events and statutory.

for the supertoken: I kinda need that in production :D as I plan to use it in events and statutory for getting the members list that the regular member don't have permissions for. totally agree on keeping it a secret, it's not for public use.

**Comment by blacksph3re** [comment]
image
I think i should see them at the commits somewhere... I know what you mean with line comments but even if I go to the files I don't see them. Have you submitted them as a review?

**Comment by serge1peshcoff** [comment]
crap, I did them just for comments, not a review. I've added it now.

File: lib/omscore_web/plugs/request_logger_plug.ex [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+  def init(opts) do
+    Keyword.get(opts, :log, :info)
+  end
+
+  def call(conn, level) do
+    start = System.monotonic_time()
+
+    Conn.register_before_send(conn, fn conn ->
+      # Logger.info(inspect(conn.assigns))
```
[the remark this answers is not in the document]
Comment (blacksph3re) [reply lib/omscore_web/plugs/request_logger_plug.ex]: The commented out line and the inspect with the keys? Sure!

File: lib/omscore_web/plugs/request_logger_plug.ex [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      Logger.log(level, fn ->
+        stop = System.monotonic_time()
+        diff = System.convert_time_unit(stop - start, :native, :microsecond)
+        status = Integer.to_string(conn.status)
+
+        [
+            conn.method, ?\s,
+            conn.request_path, ?\s,
+            status, ?\s,
+            Conn.get_resp_header(conn, "content-length"), " - ",
```
[the remark this answers is not in the document]
Comment (blacksph3re) [reply lib/omscore_web/plugs/request_logger_plug.ex]: Ah okay I didn't check it in detail

File: lib/omscore_web/plugs/request_logger_plug.ex [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            formatted_diff(diff), ", ",
+            print_user(conn)
```
[the remark this answers is not in the document]
Comment (blacksph3re) [reply lib/omscore_web/plugs/request_logger_plug.ex]: Yes I believe so.

File: lib/omscore_web/plugs/request_logger_plug.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (serge1peshcoff) [reply lib/omscore_web/plugs/request_logger_plug.ex]: yeah, this one. also I think it's worth removing the default logger, we won't need it

**Comment by serge1peshcoff** [comment]
okay, now it looks nicer, thanks ;)

@blacksph3re 2 more things to consider:
1) I am getting the deprecation warning on logging, like this (not sure if it's related):

```
oms-core-elixir_1           | warning: deprecated time unit: :seconds. A time unit should be :second, :millisecond, :microsecond, :nanosecond, or a positive integer
oms-core-elixir_1           |   (guardian) lib/guardian/token/jwt.ex:335: Guardian.Token.Jwt.set_iat/1
oms-core-elixir_1           |   (guardian) lib/guardian/token/jwt.ex:194: Guardian.Token.Jwt.build_claims/5
oms-core-elixir_1           |   (guardian) lib/guardian.ex:754: Guardian.returning_tuple/1
oms-core-elixir_1           |   (guardian) lib/guardian.ex:579: Guardian.encode_and_sign/4
oms-core-elixir_1           |
```

2) can you remove the old plug?

**Comment by serge1peshcoff** [comment]
also 2 more things:
1) can you log not a response but a request body? :D 
2) can the JSON output be pretty-formatted for readability, like I did for statutory and events?

**Comment by blacksph3re** [comment]
1.1 The warning is an error in guardian, it's fixed in their github but they haven't released it yet. It will be fixed as soon as there is a new guardian version.

1.2 Which plug do you mean as to be removed?

2.1 Ah yeah of course, somehow thought you meant the response...

2.2 I will investigate

**Comment by serge1peshcoff** [comment]
@blacksph3re thanks for your response ;)
1.2  I mean there are now 2 plugs which logs queries, first is the default one (I suggest removing it) and second is our custom one.

**Comment by serge1peshcoff** [comment]
for 2.2 this should work

`inspect(conn.resp_body, pretty: true)`

**Comment by blacksph3re** [comment]
I also added bruteforce protection to the login route, though that could maybe be a separate PR... It introduces a maximum number of login tries per 5 minutes of 100, so should not affect normal operation (except within 5 minutes 100 people with the same password log in)

**Comment by serge1peshcoff** [comment]
super awesome. one thing @blacksph3re for now the headers are quite difficult to read. can you either prettify or remove them? (not sure if we'll need it actually)

**Comment by blacksph3re** [comment]
Ay, I didn't realize I left them in there, had it for debugging the headers which traefik sets.

**Pull request merged**